### PR TITLE
WIP/RFC: Add DNS zone creation support for GCP.

### PR DESCRIPTION
**Pull request by nat-henderson**
Core question: Is this an okay direction to proceed in for getting GCP DNS integration started?  I will build the teardown / terraform code if so!

Done:
* Add GCP's DNS managed zone as a task.
* Add a DNS model for GCP, during cluster configuration lifecycle, which
    creates a zone if Cluster.Spec.DNSZone is set.

Not Done:
* Does not yet tear down the DNS managed zone or records (autodeleting
  DNS is spooky!)
* Does not yet export to Terraform
* Currently always uses private managed zones.

### CODE REVIEW

File: upup/pkg/fi/cloudup/utils.go [hunk 1]
```diff
@@ -213,12 +213,10 @@ func FindDNSHostedZone(dns dnsprovider.Interface, clusterDNSName string, dnsType
 	}
 
 	if len(maxLengthZones) == 0 {
-		// We make this an error because you have to set up DNS delegation anyway
```
[the remark this answers is not in the document]
Comment (justinsb) [reply upup/pkg/fi/cloudup/utils.go:216]: Having found cluster.spec.topoplogy.dns.type=private, I'm now thinking that's the field we should use if we can.  It looks like we don't support auto-creation on AWS though, which is surprising (I'll have to try it out!)

I propose we treat "no matching zone" as an error if dns.type is public, but we should auto-create it (as you've done here) if dns.type is private.

**Comment by justinsb** [comment]
I think this looks great, particularly if we can reuse cluster.spec.topology.dns.type so that users have to be very intentional that they want private DNS.  I'm going to try it out on AWS though, as my first reading of the code makes it look like we don't automatically create the zone on AWS, though I feel like I'm probably missing something there!

**Comment by justinsb** [comment]
I tried it out on AWS and went splunking through the old dns-zone code.  I _think_ I introduced the code that required the zone exist (way back in cae256340b0632459f04f4d7034bf5a470e4e30f), because at the time we didn't have private zones so we'd have to set up NS delegation anyway.  It looks like we never updated it for private zones.

The tricky thing we have to figure out is ownership.  I guess if we create the zone with the full cluster name then we can delete it also.  The alternative would be that multiple clusters could share a "parent zone", but in that case we wouldn't want to delete it (and would probably want it to be pre-created).

It does look like ManagedZone supports labels, so we can rely on more than just the name to know if we created it: https://cloud.google.com/dns/docs/reference/v1/managedZones#resource

I don't know of any restrictions in creating multiple non-overlapping private zones.

Is this how you imagine managed zones being used - one per cluster?

**Comment by justinsb** [comment]
I think this generally looks good; we do want to figure out ownership (i.e. should we delete).  What did you think about the suggestion of using labels @nat-henderson ?

**Comment by nat-henderson** [comment]
It sounds good to me - I'll come back around to this next week, I expect!  :)

**Comment by nat-henderson** [comment]
Hey - sorry about that, I had it working locally and thought I'd pushed it up, but must not have.  Here it is - let's see how the tests feel post-rebase...

**Comment by nat-henderson** [comment]
Great, compiles, so:

/retest

**Pull request closed without merging**